spring-shell-core/src/main/java/org/springframework/shell/command/CommandCatalog.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.shell.command;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.springframework.shell.context.InteractionMode;
+import org.springframework.shell.context.ShellContext;
+import org.springframework.util.StringUtils;
+
+/**
+ * Interface defining contract to handle existing {@link CommandRegistration}s.
+ *
+ * @author Janne Valkealahti
+ */
+public interface CommandCatalog {
+
+	/**
+	 * Register a {@link CommandRegistration}.
+	 *
+	 * @param registration the command registration
+	 */
+	void register(CommandRegistration... registration);
+
+	/**
+	 * Unregister a {@link CommandRegistration}.
+	 *
+	 * @param registration the command registration
+	 */
+	void unregister(CommandRegistration... registration);
+
+	/**
+	 * Unregister a {@link CommandRegistration} by its command name.
+	 *
+	 * @param commandName the command name
+	 */
+	void unregister(String... commandName);
+
+	/**
+	 * Gets all {@link CommandRegistration}s mapped with their names.
+	 *
+	 * @return all command registrations
+	 */
+	Map<String, CommandRegistration> getRegistrations();
+
+	/**
+	 * Interface to resolve currently existing commands. It is useful to have fully
+	 * dynamic set of commands which may exists only if some conditions in a running
+	 * shell are met. For example if shell is targeting arbitrary server environment
+	 * some commands may or may not exist depending on a runtime state.
+	 */
+	@FunctionalInterface
+	interface CommandResolver {
```

Comment:
Any reason to put these interfaces in here nested? I have found that nesting the interfaces in designs makes it harder to resonate with. 

I would strongly prefer the default impl to be outside of the interface definition as well.

Reply:
Indeed, I've had same thoughts.